File: 894-all-possible-full-binary-trees/894-all-possible-full-binary-trees.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func allPossibleFBT(n int) []*TreeNode {
    if(n % 2 == 0){
        emptySlice := make([]*TreeNode,0)
        return emptySlice
    }
    emptySlice := make([]*TreeNode,0)
    if n == 1 {
        root := &TreeNode{}
        emptySlice = append(emptySlice,root)
        return emptySlice
    }
    remaining := n-1
    for i:=1;i<remaining;i+=2 {
        leftChilds := allPossibleFBT(i)
        rightChilds := allPossibleFBT(remaining-i)
        for _,lChild := range leftChilds {
            for _,rChild := range rightChilds {
                root := &TreeNode{
                    Val: 0,
                    Left: lChild,
                    Right : rChild,
                }
                emptySlice = append(emptySlice,root)
            }
        }
    }
    return emptySlice
}

File: 894-all-possible-full-binary-trees/894-all-possible-full-binary-trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> allPossibleFBT(int n) {
        if(n%2 == 0){
            return {};
        }
        vector<TreeNode*> ans;
        if(n == 1){
            TreeNode *root = new TreeNode();
            ans.push_back(root);
            return ans;
        }
        int remaining = n-1;
        for(int i=1;i<remaining;i+=2){
            vector<TreeNode*> leftChilds = allPossibleFBT(i);
            vector<TreeNode*> rightChilds = allPossibleFBT(remaining-i);
            int m = leftChilds.size(),p = rightChilds.size();
            for(int j=0;j<m;j++){
                for(int k=0;k<p;k++){
                    TreeNode *root = new TreeNode(0,leftChilds[j],rightChilds[k]);
                    ans.push_back(root);
                }
            }
            
        }
        return ans;
    }
};